include/zephyr/drivers/i3c.h

```diff
@@ -0,0 +1,1596 @@
+	I3C_ERROR_TE_UNKNOWN,
+
+	/** No error (not official error code) */
+	I3C_ERROR_TE_NONE,
+
+	I3C_ERROR_TE_MAX = I3C_ERROR_TE_UNKNOWN,
+	I3C_ERROR_TE_INVALID,
+};
+
+/*
+ * I3C_MSG_* are I3C Message flags.
+ */
+
+/** Write message to I3C bus. */
+#define I3C_MSG_WRITE			(0U << 0U)
+
+/** Read message from I2C bus. */
+#define I3C_MSG_READ			BIT(0)
+
+/** @cond INTERNAL_HIDDEN */
+#define I3C_MSG_RW_MASK			BIT(0)
+/** @endcond  */
+
+/** Send STOP after this message. */
+#define I3C_MSG_STOP			BIT(1)
+
+/** RESTART I3C transaction for this message.
+ *
+ * @note Not all I3C drivers have or require explicit support for this
+ * feature. Some drivers require this be present on a read message
+ * that follows a write, or vice-versa.  Some drivers will merge
+ * adjacent fragments into a single transaction using this flag; some
+ * will not.
+ */
+#define I3C_MSG_RESTART			BIT(2)
+
+/** Transfer use HDR mode */
+#define I3C_MSG_HDR			BIT(3)
+
+/** I3C HDR Mode 0 */
+#define I3C_MSG_HDR_MODE0		BIT(0)
+
+/** I3C HDR Mode 1 */
+#define I3C_MSG_HDR_MODE1		BIT(1)
+
+/** I3C HDR Mode 2 */
+#define I3C_MSG_HDR_MODE2		BIT(2)
+
+/** I3C HDR Mode 3 */
+#define I3C_MSG_HDR_MODE3		BIT(3)
+
+/** I3C HDR Mode 4 */
+#define I3C_MSG_HDR_MODE4		BIT(4)
+
+/** I3C HDR Mode 5 */
+#define I3C_MSG_HDR_MODE5		BIT(5)
+
+/** I3C HDR Mode 6 */
+#define I3C_MSG_HDR_MODE6		BIT(6)
+
+/** I3C HDR Mode 7 */
+#define I3C_MSG_HDR_MODE7		BIT(7)
+
+/** I3C HDR-DDR (Double Data Rate) */
+#define I3C_MSG_HDR_DDR			I3C_MSG_HDR_MODE0
+
+/** I3C HDR-TSP (Ternary Symbol Pure-bus) */
+#define I3C_MSG_HDR_TSP			I3C_MSG_HDR_MODE1
+
+/** I3C HDR-TSL (Ternary Symbol Legacy-inclusive-bus) */
+#define I3C_MSG_HDR_TSL			I3C_MSG_HDR_MODE2
+
+/** I3C HDR-BT (Bulk Transport) */
+#define I3C_MSG_HDR_BT			I3C_MSG_HDR_MODE3
+
+/**
+ * @brief One I3C Message.
+ *
+ * This defines one I3C message to transact on the I3C bus.
+ *
+ * @note Some of the configurations supported by this API may not be
+ * supported by specific SoC I3C hardware implementations, in
+ * particular features related to bus transactions intended to read or
+ * write data from different buffers within a single transaction.
+ * Invocations of i3c_transfer() may not indicate an error when an
+ * unsupported configuration is encountered.  In some cases drivers
+ * will generate separate transactions for each message fragment, with
+ * or without presence of @ref I3C_MSG_RESTART in #flags.
+ */
+struct i3c_msg {
+	/** Data buffer in bytes */
+	uint8_t			*buf;
+
+	/** Length of buffer in bytes */
+	uint32_t		len;
+
+	/** Flags for this message */
+	uint8_t			flags;
+
+	/**
+	 * HDR mode (@c I3C_MSG_HDR_MODE*) for transfer
+	 * if @c I3C_MSG_HDR is set in @c flags.
+	 */
+	uint8_t			hdr_mode;
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * These are for internal use only, so skip these in
+ * public documentation.
+ */
+struct i3c_target_desc;
+struct i3c_i2c_device_desc;
+
+__subsystem struct i3c_driver_api {
```

Comment:
I am aware of a I3C controller where devices are not addressed using target address directly, and there is a redirection table involved. Without using the `struct i3c_i2c_device_desc` for the controller driver to store addition data to locate the entries in the redirection table, the controller driver will need to iterate the entries in the redirection table for the entry with the desired target address. This results in additional processing for each transfer that can be avoided by shifting the burden to when the driver is being written (by having `struct i3c_i2c_device_desc`).

Reply:
I'm not quite sure what you're on about here.... but I was referring to maintain compatibility with existing i2c device drivers that use the i2c_driver_api.
So that when a device driver is on that i3c bus... it can use the same i2c calls that have always existed such as the `i2c_write()` function here for example https://github.com/zephyrproject-rtos/zephyr/blob/801fe1fa06bd6f8cc2a55079411fd4feac55eff6/drivers/eeprom/eeprom_at2x.c#L334
When it defined in the i3c controller driver... it'll just look like this for that nxp..
```c
static const struct i3c_driver_api mcux_i3c_driver_api = {
	.i2c_api = {
		.transfer = mcux_i3c_i2c_transfer,
		.configure = mcux_i3c_i2c_configure,
	},
	.do_daa = mcux_i3c_do_daa,
	.do_ccc = mcux_i3c_do_ccc,

	.i3c_device_register = mcux_i3c_dev_register,
	.i3c_i2c_device_register = mcux_i3c_i2c_dev_register,

	.i3c_xfers = mcux_i3c_transfer,
};
```


Yes, I actually use that I3C controller that uses a retaining register with the target's dyanmic address... This would actually require some additional optional apis that aren't in this struct. Similar to the ones in Linux which are attach_i3c, reattach_i3c, and the detach_i3c apis (and same for i2c). If those controllers where to be implemented, then I would expect it to add those devices into it's retaining registers with i3c_device_register and i3c_i2c_device_register. I don't this this NXP controller would need some of those...

That controller does not need to iterate through those retaining registers

Reply:
That particular controller is not using simple registers but actually have an array of records on the memory of controller hardware. The idea of using the struct is simply to store some data in the `controller_priv` field for quick lookup. Note that the size of this memory (and thus number of addressable devices) is not fixed and can vary between actual implementation on hardware. The number of entries can be queried from hardware. If the driver of such controller is implemented as you say, the driver will need to have some way of pre-allocate a lookup table at build time. This will require some configuration be specified for each board or platform, and possibly resulting in some confusion during board bring-up. If we amend the drivers to use the I3C specific API, this can be done on the main Zephyr repo. This is kind of shifting the burden from downstream users to upstream driver writers. The intent is to make it easier for downstream to consume Zephyr. That's the thought process. Either way, the actual user of the devices (e.g. sensors) would be shielded from this low level details.

TBH, either way is fine to me. I don't have much attachment to the current proposed implementation. One way or the other, someone has the burden to make everything work on the final product (whether upstream or downstream, and also the driver writers). Though, one thing to consider is that, if reusing the I2C API, the device would not know whether it is connected to a I2C or I3C controller (... or that if this information is even important).

Note that if we go the reusing route, that will not be simple drop-in either. It will require some work on the devicetree side. One is the bus property. Another one is iterating children nodes, as we need to distinguish between I2C and I3C devices.

Reply:
Though, let me get the part with only I3C devices working correctly first.

Reply:
> That particular controller is not using simple registers but actually have an array of records on the memory of controller hardware. The idea of using the struct is simply to store some data in the `controller_priv` field for quick lookup. Note that the size of this memory (and thus number of addressable devices) is not fixed and can vary between actual implementation on hardware. The number of entries can be queried from hardware. If the driver of such controller is implemented as you say, the driver will need to have some way of pre-allocate a lookup table at build time. This will require some configuration be specified for each board or platform, and possibly resulting in some confusion during board bring-up. If we amend the drivers to use the I3C specific API, this can be done on the main Zephyr repo. This is kind of shifting the burden from downstream users to upstream driver writers. The intent is to make it easier for downstream to consume Zephyr. That's the thought process. Either way, the actual user of the devices (e.g. sensors) would be shielded from this low level details.

Ah, I understand what you're saying now. Yeah, for that controller, there would have to be pre-allocated lookup table and that's how we're doing it for that controller within our Zephyr fork with a `controller_priv`.

> TBH, either way is fine to me. I don't have much attachment to the current proposed implementation. One way or the other, someone has the burden to make everything work on the final product (whether upstream or downstream, and also the driver writers). Though, one thing to consider is that, if reusing the I2C API, the device would not know whether it is connected to a I2C or I3C controller (... or that if this information is even important).
> 
> Note that if we go the reusing route, that will not be simple drop-in either. It will require some work on the devicetree side. One is the bus property. Another one is iterating children nodes, as we need to distinguish between I2C and I3C devices.

I believe that we should maintain portability with drivers existing on i2c buses to stick with that i2c api and expect all i3c controllers to implement the i2c api as well. That way any existing code that does the i2c api will 'just work' on an i3c controller. I would rather see this shifted to "upstream"... because I would see that as requiring less work and would be easier to maintain IMO but I'm also curious to hear other's opinions. It would only need to be implemented within the controller code (or a common i3c header file) rather than going back and editing all existing i2c drivers and expecting all future i2c drivers to implement i3c api calls.

You're right if we go the reusing route, that would require some 'one-time' work on the bus property (like what I tried with https://github.com/zephyrproject-rtos/zephyr/pull/32730)... but once that work is completed we should be able to distinguish between I2C and I3C devices through the  DT_INST_FOREACH_CHILD macro and the DT_INST_ON_BUS macro (I think?) that could build a `struct i3c_i2c_device_desc` to be passed in to the i3c_i2c_device_register function...

Reply:
> Though, let me get the part with only I3C devices working correctly first.

yes, for sure. I'm sure there's plenty of I3C work to be done.
Meanwhile in parallel, I'll see if can get some discussion and input from other people about how I2C devices could be implemented. There would also likely be some work needed within the zephyr build system to support a bus that as compatibility with another bus.

Reply:
I'm beginning to think another option for i2c on an i3c bus would be to use a child binding on the DTS for the i2c devices to be nested within the i3c node in the device tree. Kinda like the tca954x driver..
https://github.com/zephyrproject-rtos/zephyr/blob/0ace177a32faaf9b113a87febcfbd8ae3e53c5b3/dts/bindings/i2c/ti%2Ctca954x-base.yaml#L57